api/v1alpha1/release_webhook.go

```diff
@@ -17,38 +17,109 @@ limitations under the License.
+
+	"gomodules.xyz/jsonpatch/v2"
+	admissionv1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
 )
 
-func (r *Release) SetupWebhookWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewWebhookManagedBy(mgr).
-		For(r).
-		Complete()
-}
+const authorLabel = "release.rhtap.openshift.io/author"
 
-//+kubebuilder:webhook:path=/validate-appstudio-redhat-com-v1alpha1-release,mutating=false,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=releases,verbs=create;update,versions=v1alpha1,name=vrelease.kb.io,admissionReviewVersions=v1
+// releaseHandler describes the handler for the Release low level webhook
+type releaseHandler struct {
+	Client client.Client
+}
 
-var _ webhook.Validator = &Release{}
+// +kubebuilder:webhook:path=/mutate-appstudio-redhat-com-v1alpha1-release,mutating=true,failurePolicy=fail,sideEffects=None,groups=appstudio.redhat.com,resources=releases,verbs=create;update,versions=v1alpha1,name=mrelease.kb.io,admissionReviewVersions=v1
 
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
-func (r *Release) ValidateCreate() error {
-	return nil
+// SetupReleaseLowLevelWebhook registers the Release low level webhook with the passed manager.
+func SetupReleaseLowlevelWebhook(mgr ctrl.Manager) {
+	mgr.GetWebhookServer().Register("/mutate-appstudio-redhat-com-v1alpha1-release", &webhook.Admission{Handler: &releaseHandler{Client: mgr.GetClient()}})
 }
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
-func (r *Release) ValidateUpdate(old runtime.Object) error {
-	if !reflect.DeepEqual(r.Spec, old.(*Release).Spec) {
-		return fmt.Errorf("release resources spec cannot be updated")
+// Handle takes an incoming admission request and returns an admission response. In the case of create or update
+// requests, the response will patch the custom resource to include the author as a label. All other requests
+// are accepted without action.
+func (r *releaseHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	switch req.AdmissionRequest.Operation {
+	case admissionv1.Create:
+		return patchAuthorResponse(req)
+	case admissionv1.Update:
+		var releaseOld, releaseNew Release
+		err := json.Unmarshal(req.Object.Raw, &releaseNew)
+		if err != nil {
+			failedResponse(req.UID, "could not unmarshal the raw incoming Release object")
+		}
+		err = json.Unmarshal(req.OldObject.Raw, &releaseOld)
+		if err != nil {
+			failedResponse(req.UID, "could not unmarshal the raw old Release object")
+		}
+		if !reflect.DeepEqual(releaseNew.Spec, releaseOld.Spec) {
+			return failedResponse(req.UID, "release resources spec cannot be updated")
+		}
+		return patchAuthorResponse(req)
```

Comment:
Open question here about if (as is currently written) any update to metadata is allowed, but you become the marked author OR any update to metadata is allowed EXCEPT updates to the author label, which would cause your request to be rejected

Reply:
Discussed on slack. Resolving